cmd/mockserver/server.go

```diff
@@ -195,50 +238,57 @@ func (ds *fakeDebugSession) writeAndLogProtocolMessage(message dap.Message) {
 
 // The debugging session will keep track of how many breakpoints
 // have been set. Once start-up is done (i.e. configurationDone
-// request is processed), it will "stop" at each breakpoint one
-// by one, and once there are no more, it will trigger a terminate
-// event.
+// request is processed), it will "stop" at each breakpoint one by
+// one, and once there are no more, it will trigger a terminated event.
 type fakeDebugSession struct {
+	// rw is used to read requests and write events/responses
 	rw *bufio.ReadWriter
 
-	// breakpointSet is a counter of the remaining breakpoints
-	// that the debug session is yet to stop at before the program
-	// terminates. It must be 0 at start-up and termination.
-	breakpointsSet int
-
-	// canContinue is used to implement a small state machine between
-	// multiple server functions. The debug session is paused
-	// (canContinue is false) while multiple client requests are being
-	// processed during the start-up sequence or when stopping at a
-	// breakpoint. Once the client allows the session to continue,
-	// the value is flipped to true. It is reset back to false
-	// at termination and is ready for the next client session.
-	canContinue bool
-}
-
-// doContinue is to be called between handling of each request/response
-// to simulate events from the debug session that is in continue mode.
-// If the program is "stopped", this will be a no-op. Otherwise, this
-// will "stop" on a breakpoint or terminate if there are no more
-// breakpoints.
+	// sendQueue is used to capture messages from multiple request
+	// processing goroutines while writing them to the client connection
+	// from a single goroutine via sendFromQueue. We must keep track of
+	// the multiple channel senders with a wait group to make sure we do
+	// not close this channel prematurely. Closing this channel will signal
+	// the sendFromQueue goroutine that it can exit.
+	sendQueue chan dap.Message
+	sendWg    sync.WaitGroup
+
+	// stopDebug is used notify long-running handlers to stop processing.
+	stopDebug chan struct{}
+
+	// bpSet is a counter of the remaining breakpoints that the debug
+	// session is yet to stop at before the program terminates.
+	bpSet    int
+	bpSetMux sync.Mutex
+}
+
+// doContinue allows fake program execution to continue when the program
+// is started or unpaused. It simulates events from the debug session
+// by "stopping" on a breakpoint or terminating if there are no more
+// breakpoints. Safe to use concurrently.
 func (ds *fakeDebugSession) doContinue() {
-	if !ds.canContinue {
-		return
-	}
-	ds.canContinue = false
 	var e dap.Message
-	if ds.breakpointsSet == 0 {
-		e = &dap.TerminatedEvent{
-			Event: *newEvent("terminated"),
-		}
-	} else {
-		e = &dap.StoppedEvent{
-			Event: *newEvent("stopped"),
-			Body:  dap.StoppedEventBody{Reason: "breakpoint", ThreadId: 1, AllThreadsStopped: true},
+	func() {
+		ds.bpSetMux.Lock()
```

Comment:
I'm not sure you need the func() here, since there's no loop or anything. Is it for the mutex scoping? It should be alright without it too

there's no returns here so you can just unlock manually before `ds.send` ?

Reply:
Removed. It is no longer needed. There used to be an early return that required it, but I factored it out.